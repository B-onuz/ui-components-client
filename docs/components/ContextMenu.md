# ContextMenu (Component)

Componente de menu de ações vinculados a um button

## Como utilizar

```js
import ContextMenu from 'ui-components/lib/components/ContextMenu'
```

```jsx
import React from 'react'
import ContextMenu from 'ui-components/lib/components/ContextMenu'
import { Flex, Box } from 'reflexbox/styled-components'

const functionA = () => {
  console.log('Função A')
}

const functionB = () => {
  console.log('Função B')
}

const functionC = () => {
  console.log('Função C')
}

const exampleObj = [
  {
    text: 'editar',
    buttonFunction: functionA,
  },
  {
    text: 'adicionar',
    buttonFunction: functionB,
  },
  {
    text: 'excluir',
    buttonFunction: functionC,
  },
]

export const ContextMenuExample = () => (
  <Flex style={{ height: '100vh' }} width={[1]} justifyContent={'space-evenly'} alignItems={'center'}>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'right'} />
    </Box>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'top'} />
    </Box>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'left'} />
    </Box>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'bottom'} />
    </Box>
  </Flex>
)
```
