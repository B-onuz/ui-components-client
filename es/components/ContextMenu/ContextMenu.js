"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styledSystem = require("styled-system");

var _reactFontawesome = require("@fortawesome/react-fontawesome");

var _freeSolidSvgIcons = require("@fortawesome/free-solid-svg-icons");

var _Button = _interopRequireDefault(require("../Button"));

var _ContextMenuDialog = _interopRequireDefault(require("./ContextMenuDialog"));

var _Modal = _interopRequireDefault(require("../Modal"));

var _useModal2 = _interopRequireDefault(require("../../hooks/useModal"));

var _shortid = _interopRequireDefault(require("shortid"));

var _this = void 0,
    _jsxFileName = "/Users/master/projects/bonuz/ui-components/src/components/ContextMenu/ContextMenu.js";

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject5() {
  var data = _taggedTemplateLiteralLoose(["\n  color: ", ";\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose(["\n  width: 100%;\n  margin: 8px 0;\n  border: 1px solid #f1f1f3;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n  width: 2em;\n  height: 2em;\n  border-radius: 50%;\n  background-color: #f5f5f5;\n  box-shadow: 0px 2px 3px #2c282848;\n  border: none;\n  &:hover {\n    cursor: pointer;\n  }\n  &:focus {\n    outline: none;\n  }\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n  display: flex;\n  flex-direction: column;\n  padding: 20px 20px;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  ", "\n  ", "\n  position: relative;\n  display: flex;\n  .contextButton {\n    width: 2em;\n    height: 2em;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteralLoose(strings, raw) { if (!raw) { raw = strings.slice(0); } strings.raw = raw; return strings; }

var StyledContextMenu = _styledComponents["default"].div(_templateObject(), _styledSystem.color, _styledSystem.space);

var ModalWrapper = _styledComponents["default"].div(_templateObject2());

var ContextMenuToogleButton = _styledComponents["default"].button(_templateObject3());

var Divider = _styledComponents["default"].hr(_templateObject4());

var Icon = (0, _styledComponents["default"])(_reactFontawesome.FontAwesomeIcon)(_templateObject5(), function (_ref) {
  var theme = _ref.theme,
      buttonIcon = _ref.buttonIcon;
  return !buttonIcon ? theme.colors.darkGrey : '';
});

var ContextMenu = function ContextMenu(_ref2) {
  var children = _ref2.children,
      direction = _ref2.direction,
      contextMenuActions = _ref2.contextMenuActions,
      contextFunctions = _ref2.contextFunctions,
      buttonIcon = _ref2.buttonIcon,
      rest = _objectWithoutPropertiesLoose(_ref2, ["children", "direction", "contextMenuActions", "contextFunctions", "buttonIcon"]);

  var _useState = (0, _react.useState)(false),
      isOpen = _useState[0],
      setIsOpen = _useState[1];

  var wrapperRef = (0, _react.useRef)(null);
  var buttonRef = (0, _react.useRef)(null);

  var _useModal = (0, _useModal2["default"])(),
      isOpenResponsiveModal = _useModal.isOpen,
      openResponsiveModal = _useModal.openModal,
      closeResponsiveModal = _useModal.closeModal,
      responsiveModalParams = _useModal.params;

  var onClose = function onClose() {
    setIsOpen(false);
    closeResponsiveModal();
    return true;
  };

  (0, _react.useEffect)(function () {
    if (isOpen) {
      openResponsiveModal();
    }
  });
  var needResponsive = screen.width >= 500 ? false : true;
  return /*#__PURE__*/_react["default"].createElement(StyledContextMenu, _extends({
    ref: wrapperRef,
    key: "" + _shortid["default"].generate()
  }, rest, {
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 81,
      columnNumber: 5
    }
  }), contextFunctions ? /*#__PURE__*/_react["default"].createElement("div", {
    ref: buttonRef,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 83,
      columnNumber: 9
    }
  }, /*#__PURE__*/_react["default"].createElement(_Button["default"], {
    p: 0,
    className: "contextButton",
    onClick: function onClick() {
      return setIsOpen(function (last) {
        return !last;
      });
    },
    color: 'primary',
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 84,
      columnNumber: 11
    }
  }, /*#__PURE__*/_react["default"].createElement(Icon, {
    buttonIcon: true,
    icon: buttonIcon,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 85,
      columnNumber: 13
    }
  }))) : /*#__PURE__*/_react["default"].createElement(ContextMenuToogleButton, {
    ref: buttonRef,
    onClick: function onClick() {
      return setIsOpen(function (last) {
        return !last;
      });
    },
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 89,
      columnNumber: 9
    }
  }, /*#__PURE__*/_react["default"].createElement(Icon, {
    icon: _freeSolidSvgIcons.faEllipsisH,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 90,
      columnNumber: 11
    }
  })), needResponsive ? /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(_Modal["default"], {
    open: isOpenResponsiveModal,
    onClose: onClose,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 95,
      columnNumber: 11
    }
  }, /*#__PURE__*/_react["default"].createElement(ModalWrapper, {
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 96,
      columnNumber: 13
    }
  }, (contextMenuActions || []).map(function (action, index) {
    return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, {
      key: action.text + "-" + _shortid["default"].generate(),
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 98,
        columnNumber: 17
      }
    }, /*#__PURE__*/_react["default"].createElement(_Button["default"], {
      key: action.text + "-" + _shortid["default"].generate(),
      linkButton: true,
      color: 'primary',
      onClick: function onClick() {
        return onClose() && action.buttonFunction();
      },
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 99,
        columnNumber: 19
      }
    }, action.text), index + 1 !== contextMenuActions.length ? /*#__PURE__*/_react["default"].createElement(Divider, {
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 107,
        columnNumber: 62
      }
    }) : null);
  })))) : /*#__PURE__*/_react["default"].createElement(_ContextMenuDialog["default"], {
    contextFunctions: contextFunctions,
    isOpen: isOpen,
    onClose: onClose,
    buttonRef: buttonRef,
    direction: direction,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 114,
      columnNumber: 9
    }
  }, (contextMenuActions || []).map(function (action, index) {
    return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, {
      key: action.text + "-" + _shortid["default"].generate(),
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 116,
        columnNumber: 13
      }
    }, /*#__PURE__*/_react["default"].createElement(_Button["default"], {
      key: action.text + "-" + _shortid["default"].generate(),
      linkButton: true,
      color: 'primary',
      onClick: function onClick() {
        return onClose() && action.buttonFunction();
      },
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 117,
        columnNumber: 15
      }
    }, action.text), index + 1 !== contextMenuActions.length ? /*#__PURE__*/_react["default"].createElement(Divider, {
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 120,
        columnNumber: 58
      }
    }) : null);
  })));
};

ContextMenu.defaultProps = {
  direction: 'left'
};
var _default = ContextMenu;
exports["default"] = _default;