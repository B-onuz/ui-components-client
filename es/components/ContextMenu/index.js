"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ContextMenu = _interopRequireDefault(require("./ContextMenu"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _default = _ContextMenu["default"];
exports["default"] = _default;