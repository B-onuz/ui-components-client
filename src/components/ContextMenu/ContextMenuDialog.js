import React, { useState, useEffect } from 'react'
import styled from 'styled-components'
import ReactDOM from 'react-dom'

const Dialog = styled.div`
  position: absolute;
  border-radius: 12px;
  opacity: ${({ isOpen }) => (isOpen ? 1 : 0)};
  z-index: ${({ isOpen }) => (isOpen ? 1 : -100)};
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 10px 16px;
  bottom: ${({ bottom }) => bottom};
  top: ${({ top }) => top};
  left: ${({ left }) => left};
  background: #fff 0% 0% no-repeat padding-box;
  box-shadow: 0px 2px 6px #2c28281c;
  transform: ${({ direction, bottom }) => {
    if (direction === 'left') {
      return `
        translateX(calc(-100% - 32px * 2  - 5px))
      `
    }
    if (direction === 'right') {
      return `
        none
      `
    }
    if (direction === 'top') {
      return `
        translateX(calc(-50% - 32px / 2)) translateY(calc(-100% - 24px));
      `
    }
    if (direction === 'bottom' && bottom === 'auto') {
      return `
      translateX(calc(-50% - 32px / 2)) translateY(calc(-100% - 24px));
      `
    } else {
      return `
        translateX(calc(-50% - 32px / 2));
      `
    }
  }};
  &::after {
    content: '';
    width: 1em;
    height: 1em;
    position: absolute;
    box-shadow: 1px 2px 2px -1px #2c28281c;
    background: #fff 0% 0%;
    top: ${({ top }) => (top !== 'unset' ? '1.6em' : 'unset')};
    bottom: ${({ bottom }) => (bottom !== 'unset' ? '1.6em' : 'unset')};
    ${({ direction, bottom }) => {
      if (direction === 'left') {
        return `
      right: -0.5em;
      transform: rotateZ(-45deg);
    `
      }
      if (direction === 'right') {
        return `
      left: -0.5em;
      transform: rotateZ(135deg);
    `
      }
      if (direction === 'top') {
        return `
      top: calc(100% - 0.5em);
      transform: rotateZ(45deg);
      left: 0;
      right: 0;
      margin: auto;
    `
      }
      if (direction === 'bottom' && bottom === 'auto') {
        return `
          top: calc(100% + 1.5em);
          transform: rotateZ(45deg);
          left: 0;
          right: 0;
          margin: auto;
        `
      } else {
        return `
          bottom: calc(100% + 1.5em);
          transform: rotateZ(-135deg);
          left: 0;
          right: 0;
          margin: auto;
        `
      }
    }}
  }

  @media only screen and (max-width: 500px) {
    position: fixed;
    margin: 0;
    top: 50%;
    left: 50%;
    margin-right: -50%;
    transform: translate(-50%, -50%);
    padding: 20px 30px;

    &::after {
      display: none;
    }
  }
`

const useRootDialog = () => {
  const [$rootDialog, setRootDialog] = useState()
  const IdName = 'contextMenu-root'

  useEffect(() => {
    const rootModal = document.getElementById(IdName)
    if (rootModal) {
      setRootDialog(rootModal)
    } else {
      const rootEl = document.createElement('div')
      rootEl.id = IdName
      setRootDialog(rootEl)
      document.body.appendChild(rootEl)
    }
  })

  return { $rootDialog }
}

const Wrapper = styled.div`
  z-index: 99999999999;
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
`

const ContextMenuDialog = ({ isOpen, buttonRef, onClose, direction, contextFunctions, ...rest }) => {
  const { $rootDialog } = useRootDialog()
  const screenHeight = window.innerHeight
  const $currentButtonRef = (buttonRef || {}).current
  const [top, setTop] = useState(0)
  const [left, setLeft] = useState(0)
  const [bottom, setBottom] = useState(0)

  const updatePositions = () => {
    if (!$currentButtonRef) return null
    const offset = $currentButtonRef.getBoundingClientRect()

    if (direction === 'top') {
      setLeft(`${offset.left + offset.width}px`)
      setTop(`${offset.top + offset.height - 24}px`)
      setBottom('unset')
    }

    if (offset.top < screenHeight / 2) {
      if (direction === 'right') {
        setLeft(`${offset.left + offset.width + 20}px`)
        setTop(`${offset.top - offset.height / 2}px`)
        setBottom('unset')
      }

      if (direction === 'left') {
        setTop(`${offset.top - offset.height / 2}px`)
        setLeft(`${offset.left + offset.width + 20}px`)
        setBottom('unset')
      }

      if (direction === 'bottom') {
        setTop(`${offset.top + offset.height + 24}px`)
        setLeft(`${offset.left + offset.width}px`)
        setBottom('unset')
      }
    } else {
      if (direction === 'right') {
        setBottom(`${screenHeight - offset.bottom - 16}px`)
        setLeft(`${offset.left + offset.width + 20}px`)
        setTop('unset')
      }

      if (direction === 'left') {
        setBottom(`${screenHeight - offset.bottom - 16}px`)
        setLeft(`${offset.left + offset.width + 20}px`)
        setTop('unset')
      }

      if (direction === 'bottom') {
        setLeft(`${offset.left + offset.width}px`)
        setTop(`${offset.top + offset.height - 24}px`)
        setBottom('auto')
      }
    }
  }

  useEffect(() => {
    updatePositions()
    window.addEventListener('resize', updatePositions)
    return () => {
      window.removeEventListener('resize', updatePositions)
    }
  })

  if (!$rootDialog) return null

  if (!isOpen) {
    return null
  }

  return ReactDOM.createPortal(
    <Wrapper onClick={onClose}>
      <Dialog direction={direction} isOpen={isOpen} top={top} left={left} bottom={bottom} {...rest} contextFunctions={contextFunctions} />
    </Wrapper>,
    $rootDialog
  )
}

export default ContextMenuDialog
