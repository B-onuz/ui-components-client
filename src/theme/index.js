const colors = {
  red: '#db2828',
  lightRed: '#ef4d5a',
  mediumRed: '#FF4F5D',
  darkRed: '#9c2543',
  fullDarkRed: '#550e33',
  orange: '#f2711c',
  yellow: '#fbbd08',
  olive: '#b5cc18',
  green: '#21BA45',
  teal: '#00b5ad',
  blue: '#2185d0',
  facebookBlue: '#3B5998',
  blueLink: '#4285F4',
  violet: '#6435c9',
  purple: '#a333c8',
  pink: '#e03997',
  brown: '#a5673f',
  grey: '#767676',
  textGrey: '#8D8D8D',
  lightGrey: '#6D6D6D',
  mediumGrey: '#838383',
  darkGrey: '#9A9A9A',
  boldGray: '#55555A',
  black: '#1b1c1d',
  white: '#fdfdfd',
  warning: '#7A4D05',
  success: '#1E561F',
  info: '#0E566C',
  danger: '#973937',
  cyan: '#1CB5AD',
  disabledGrey: '#ffffff1f',
  disabledGreyText: '#ffffff4d',
}

colors.primary = colors.lightRed
colors.secondary = colors.fullDarkRed
colors.default = '#e0e1e2'

const breakpoints = ['40em', '52em', '64em', '80em']

breakpoints.sm = breakpoints[0]
breakpoints.md = breakpoints[1]
breakpoints.lg = breakpoints[2]
breakpoints.xl = breakpoints[3]

const fontSizes = [12, 14, 16, 20, 24, 32]

export default {
  fontFamily: `'Lato', sans-serif`,
  colors,
  fontSizes,
  breakpoints,
}
