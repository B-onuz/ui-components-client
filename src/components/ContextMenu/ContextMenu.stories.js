import React from 'react'
import ContextMenu from './ContextMenu'
import { Flex, Box } from 'reflexbox/styled-components'
import { action } from '@storybook/addon-actions'
import { faPlus, faCloudUploadAlt } from '@fortawesome/free-solid-svg-icons'

export default {
  title: 'ContextMenu',
  component: ContextMenu,
}

const functionA = action('Função A')

const functionB = action('Função B')

const functionC = action('Função C')

const exampleObj = [
  {
    text: 'editar',
    buttonFunction: functionA,
  },
  {
    text: 'adicionar',
    buttonFunction: functionB,
  },
  {
    text: 'excluir',
    buttonFunction: functionC,
  },
]

const singleObj = [
  {
    text: 'editar',
    buttonFunction: functionA,
  },
]

const contextOptions = [
  {
    text: 'editar',
    buttonFunction: functionA,
  },
  {
    text: 'adicionar',
    buttonFunction: functionB,
  },
  {
    text: 'excluir',
    buttonFunction: functionC,
  },
]

export const ContextMenuExample = () => (
  <>
    <Flex
      style={{ height: '100px', border: '1px solid #f0f', marginTop: '40px', overflow: 'hidden' }}
      width={[1]}
      justifyContent={'space-evenly'}
      alignItems={'center'}
    >
      <Box p={4}>
        <ContextMenu contextMenuActions={singleObj} direction={'right'} />
      </Box>
      <Box p={4}>
        <ContextMenu contextMenuActions={singleObj} direction={'top'} />
      </Box>
      <Box p={4}>
        <ContextMenu contextMenuActions={singleObj} direction={'left'} />
      </Box>
      <Box p={4}>
        <ContextMenu contextMenuActions={singleObj} direction={'bottom'} />
      </Box>
    </Flex>
    <Flex
      style={{ height: '100px', border: '1px solid #f0f', marginTop: '40px', overflow: 'hidden' }}
      width={[1]}
      justifyContent={'space-evenly'}
      alignItems={'center'}
    >
      <Box p={4}>
        <ContextMenu contextMenuActions={exampleObj} direction={'right'} />
      </Box>
      <Box p={4}>
        <ContextMenu contextMenuActions={exampleObj} direction={'top'} />
      </Box>
      <Box p={4}>
        <ContextMenu contextMenuActions={exampleObj} direction={'left'} />
      </Box>
      <Box p={4}>
        <ContextMenu contextMenuActions={exampleObj} direction={'bottom'} />
      </Box>
    </Flex>
    <Flex
      style={{ height: '100px', border: '1px solid #f0f', marginTop: '40px', overflow: 'hidden' }}
      width={[1]}
      justifyContent={'space-evenly'}
      alignItems={'center'}
    >
      <Box p={4}>
        <ContextMenu contextMenuActions={contextOptions} contextFunctions buttonIcon={faPlus} direction={'right'} />
      </Box>
      <Box p={4}>
        <ContextMenu contextMenuActions={contextOptions} contextFunctions buttonIcon={faPlus} direction={'left'} />
      </Box>
      <Box p={4}>
        <ContextMenu contextMenuActions={contextOptions} contextFunctions buttonIcon={faPlus} direction={'top'} />
      </Box>
      <Box p={4}>
        <ContextMenu contextMenuActions={contextOptions} contextFunctions buttonIcon={faPlus} direction={'bottom'} />
      </Box>
    </Flex>
  </>
)
