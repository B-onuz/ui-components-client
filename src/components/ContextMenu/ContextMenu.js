import React, { useRef, useState, useEffect } from 'react'
import styled from 'styled-components'
import { color, space } from 'styled-system'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faEllipsisH } from '@fortawesome/free-solid-svg-icons'
import Button from '../Button'
import ContextMenuDialog from './ContextMenuDialog'
import Modal from '../Modal'
import useModal from '../../hooks/useModal'
import shortid from 'shortid'

const StyledContextMenu = styled.div`
  ${color}
  ${space}
  position: relative;
  display: flex;
  .contextButton {
    width: 2em;
    height: 2em;
  }
`

const ModalWrapper = styled.div`
  display: flex;
  flex-direction: column;
  padding: 20px 20px;
`

const ContextMenuToogleButton = styled.button`
  width: 2em;
  height: 2em;
  border-radius: 50%;
  background-color: #f5f5f5;
  box-shadow: 0px 2px 3px #2c282848;
  border: none;
  &:hover {
    cursor: pointer;
  }
  &:focus {
    outline: none;
  }
`

const Divider = styled.hr`
  width: 100%;
  margin: 8px 0;
  border: 1px solid #f1f1f3;
`

const Icon = styled(FontAwesomeIcon)`
  color: ${({ theme, buttonIcon }) => (!buttonIcon ? theme.colors.darkGrey : '')};
`

const ContextMenu = ({ children, direction, contextMenuActions, contextFunctions, buttonIcon, ...rest }) => {
  const [isOpen, setIsOpen] = useState(false)
  const wrapperRef = useRef(null)
  const buttonRef = useRef(null)

  const {
    isOpen: isOpenResponsiveModal,
    openModal: openResponsiveModal,
    closeModal: closeResponsiveModal,
    params: responsiveModalParams,
  } = useModal()

  const onClose = () => {
    setIsOpen(false)
    closeResponsiveModal()
    return true
  }

  useEffect(() => {
    if (isOpen) {
      openResponsiveModal()
    }
  })

  const needResponsive = screen.width >= 500 ? false : true

  return (
    <StyledContextMenu ref={wrapperRef} key={`${shortid.generate()}`} {...rest}>
      {contextFunctions ? (
        <div ref={buttonRef}>
          <Button p={0} className="contextButton" onClick={() => setIsOpen((last) => !last)} color={'primary'}>
            <Icon buttonIcon icon={buttonIcon} />
          </Button>
        </div>
      ) : (
        <ContextMenuToogleButton ref={buttonRef} onClick={() => setIsOpen((last) => !last)}>
          <Icon icon={faEllipsisH} />
        </ContextMenuToogleButton>
      )}
      {needResponsive ? (
        <>
          <Modal open={isOpenResponsiveModal} onClose={onClose}>
            <ModalWrapper>
              {(contextMenuActions || []).map((action, index) => (
                <React.Fragment key={`${action.text}-${shortid.generate()}`}>
                  <Button
                    key={`${action.text}-${shortid.generate()}`}
                    linkButton
                    color={'primary'}
                    onClick={() => onClose() && action.buttonFunction()}
                  >
                    {action.text}
                  </Button>
                  {index + 1 !== contextMenuActions.length ? <Divider /> : null}
                </React.Fragment>
              ))}
            </ModalWrapper>
          </Modal>
        </>
      ) : (
        <ContextMenuDialog contextFunctions={contextFunctions} isOpen={isOpen} onClose={onClose} buttonRef={buttonRef} direction={direction}>
          {(contextMenuActions || []).map((action, index) => (
            <React.Fragment key={`${action.text}-${shortid.generate()}`}>
              <Button key={`${action.text}-${shortid.generate()}`} linkButton color={'primary'} onClick={() => onClose() && action.buttonFunction()}>
                {action.text}
              </Button>
              {index + 1 !== contextMenuActions.length ? <Divider /> : null}
            </React.Fragment>
          ))}
        </ContextMenuDialog>
      )}
    </StyledContextMenu>
  )
}

ContextMenu.defaultProps = {
  direction: 'left',
}

export default ContextMenu
