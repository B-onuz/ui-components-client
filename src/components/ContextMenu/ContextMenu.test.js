import React from 'react'
import ContextMenu from './ContextMenu.js'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import theme from '../../theme'
import 'jest-styled-components'

test('Should render ContextMenu', () => {
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <ContextMenu />
    </ThemeProvider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
