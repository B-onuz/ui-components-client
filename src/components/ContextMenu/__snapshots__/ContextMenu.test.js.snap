// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ContextMenu 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 .contextButton {
  width: 2em;
  height: 2em;
}

.c1 {
  width: 2em;
  height: 2em;
  border-radius: 50%;
  background-color: #f5f5f5;
  box-shadow: 0px 2px 3px #2c282848;
  border: none;
}

.c1:hover {
  cursor: pointer;
}

.c1:focus {
  outline: none;
}

.c2 {
  color: #9A9A9A;
}

<div
  className="c0"
>
  <button
    className="c1"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-ellipsis-h fa-w-16 c2"
      data-icon="ellipsis-h"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M328 256c0 39.8-32.2 72-72 72s-72-32.2-72-72 32.2-72 72-72 72 32.2 72 72zm104-72c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72zm-352 0c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
</div>
`;
